src/query/functions/binary/binary.go

```diff
@@ -127,98 +142,47 @@ var (
 	errNoMatching              = errors.New("vector matching parameters must be provided for binary operations between series")
 )
 
-type binaryOp struct {
-	OperatorType string
-	fn           mathFunc
-	params       NodeParams
-}
-
-// OpType for the operator
-func (o binaryOp) OpType() string {
-	return o.OperatorType
-}
-
-// String representation
-func (o binaryOp) String() string {
-	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.params.LNode, o.params.RNode)
-}
-
-// Node creates an execution node
-func (o binaryOp) Node(controller *transform.Controller) transform.OpNode {
-	return &binaryNode{
-		controller: controller,
-		cache:      transform.NewBlockCache(),
-		op:         o,
-	}
-}
-
-type binaryNode struct {
-	op         binaryOp
-	controller *transform.Controller
-	cache      *transform.BlockCache
-	mu         sync.Mutex
-}
-
-// NodeParams describes the types of nodes used
-// for binary operations
-type NodeParams struct {
-	LNode, RNode         parser.NodeID
-	LIsScalar, RIsScalar bool
-	ReturnBool           bool
-	VectorMatching       *logical.VectorMatching
-}
-
-// NewBinaryOp creates a new binary operation
-func NewBinaryOp(
+// Builds a logical processing function if able. If wrong opType supplied,
+// returns no function and false
+func buildLogicalFunction(
 	opType string,
 	params NodeParams,
-) (parser.Params, error) {
+) (processFunc, bool) {
+	if params.ReturnBool {
+		opType += returnBoolSuffix
+	}
 	fn, ok := mathFuncs[opType]
 	if !ok {
-		return binaryOp{}, fmt.Errorf("operator not supported: %s", opType)
+		return nil, false
 	}
 
-	return binaryOp{
-		OperatorType: opType,
-		fn:           fn,
-		params:       params,
-	}, nil
+	return createBinaryProcessingStep(params, fn), true
 }
 
-// Process processes a block
-func (n *binaryNode) Process(ID parser.NodeID, b block.Block) error {
-	lhs, rhs, err := n.computeOrCache(ID, b)
-	if err != nil {
-		// Clean up any blocks from cache
-		n.cleanup()
-		return err
-	}
-
-	// Both blocks are not ready
-	if lhs == nil || rhs == nil {
-		return nil
-	}
-
-	n.cleanup()
-
-	nextBlock, err := n.process(lhs, rhs)
-	if err != nil {
-		return err
+// Build the binary processing step
+func createBinaryProcessingStep(
+	params NodeParams,
+	fn mathFunc,
+) func(block.Block, block.Block, *transform.Controller) (block.Block, error) {
+	return func(lhs, rhs block.Block, controller *transform.Controller) (block.Block, error) {
+		return processBinary(lhs, rhs, params, controller, fn)
 	}
-
-	defer nextBlock.Close()
-	return n.controller.Process(nextBlock)
 }
 
 // processes two logical blocks, performing a logical operation on them
-func (n *binaryNode) process(lhs, rhs block.Block) (block.Block, error) {
+func processBinary(
```

Comment:
the signature could become simpler if this was defined on the node object as before.

Reply:
I'd prefer a more complex function generator signature in exchange for a simpler node

Reply:
but then you're passing the same arguments multiple layers deep. This provides an interesting comparison: https://stackoverflow.com/questions/36904955/passing-single-object-vs-passing-multiple-parameters